### DISCUSSION TITLE
[Proposal][NotNull Delegates and Events]

### DISCUSSION THREAD

**Discussion by redradist** [dotnet/csharplang#2159] [Language Ideas] [answered]
C# 8 introduces NotNull Reference Types
Let's make "surrogate reference" (events, delegate) also compatible to this feature.

My proposal instead of doing the following:

```csharp
using System;

namespace SimpleConsoleApp
{
    class Client
    {
        public Client(Server server)
        {
            server.ServerActionEvent += OnServerActionEvent;
        }
        
        public void OnServerActionEvent(int eventId)
        {
            Console.WriteLine($"OnServerActionEvent(int eventId[{eventId}])");
        }
    }

    class Server
    {
        public event Action<int> ServerActionEvent;

        public void DoSomeAction()
        {
            ServerActionEvent?.Invoke(1);
        }

    }
    
    class Program
    {
    
        static void Main(string[] args)
        {
            Server server = new Server();
            Client client = new Client(server);
            server.DoSomeAction();
        }
    }
}
```

Lets introduce implicit **Not Null Events**:

```csharp
using System;

namespace SimpleConsoleApp
{
    class Client
    {
        public Client(Server server)
        {
            server.ServerActionEvent += OnServerActionEvent;
        }
        
        public void OnServerActionEvent(int eventId)
        {
            Console.WriteLine($"OnServerActionEvent(int eventId[{eventId}])");
        }
    }

    class Server
    {
        public event Action<int> ServerActionEvent;

        public void DoSomeAction()
        {
            ServerActionEvent(1);
            // Under the hood it will call the following code as previously
            // ServerActionEvent?.Invoke(1);
        }

    }
    
    class Program
    {
    
        static void Main(string[] args)
        {
            Server server = new Server();
            Client client = new Client(server);
            server.DoSomeAction();
        }
    }
}
```

The same for delegates (**Not Null Delegates**)

What do you think guys ?
Share you opinion ...

**Comment by HaloFour** [comment]
It looks like you're proposing silently changing the behavior of existing legal syntax.

**Comment by redradist** [comment]
@HaloFour 
> 
> 
> It looks like you're proposing silently changing the behavior of existing legal syntax.

Yes, but we try to fix the billion dollar mistake, don't we ? If we will change syntax a little bit who lose ?

**Comment by svick** [comment, marked as answer]
> If we will change syntax a little bit who lose ?

Everyone who sees the new code, uses it with an old compiler and gets a ticking bomb in their code. With any other feature, attempting to use it with an old compiler will result in a compile-time error, which is a much better experience.

> we try to fix the billion dollar mistake, don't we ? 

Yes, but we're doing it very carefully, and we're not actually changing the behavior of existing programs (only which warnings they produce and even then, only if you opt-in).
